feat(mcp, firestore): Add 'select' parameter to get_documents tool

boraberkcetintas opened the pull request:
### Description

Adds a 'select' parameter to the `firestore_get_documents` MCP tool. This allows developers to specify which fields of a document to return, preventing token limit errors when dealing with large documents.

The `select` parameter accepts an array of field paths (e.g., ['field1', 'nested.field']). The MCP server will still fetch the full document from Firestore but will apply this field mask before sending the response to the LLM.

If `select` is not provided or is an empty array, the full document is returned. The `__path__` field is always preserved in the response for context.

### Scenarios Tested

A comprehensive suite of unit tests has been added to cover the new functionality.

### Sample Commands

firebase firestore documents get "my-collection/my-doc" --select "field1,nested.field"

Fixes  #9043

boraberkcetintas committed:
feat(mcp, firestore): Add 'select' parameter to get_documents tool

### Description

Adds a 'select' parameter to the `firestore_get_documents` MCP tool. This allows developers to specify which fields of a document to return, preventing token limit errors when dealing with large documents.

The `select` parameter accepts an array of field paths (e.g., ['field1', 'nested.field']). The MCP server will still fetch the full document from Firestore but will apply this field mask before sending the response to the LLM.

If `select` is not provided or is an empty array, the full document is returned. The `__path__` field is always preserved in the response for context.

### Scenarios Tested

A comprehensive suite of unit tests has been added to cover the new functionality.

### Sample Commands

firebase firestore documents get "my-collection/my-doc" --select "field1,nested.field"

Fixes #9043

joehan reviewed (approved):
Code LGTM, but gonna have someone from the Firestore team take a look as well - the Gemini suggested change seems like it might be a good idea, but IDK if it is a hallucination or not.

boraberkcetintas committed:
Update src/mcp/tools/firestore/get_documents.ts

Co-authored-by: Joe Hanley <joehanley@google.com>

boraberkcetintas committed:
Merge branch 'master' into feat-mcp-firestore-select-field

boraberkcetintas commented:
@joehan Gemini’s idea seems fair. I developed the suggested method, but I couldn’t figure out how to use the Firestore Emulator and MCP together to test it myself. Since the Admin SDK has a select feature, I think this approach could work.

I also built the package locally with npm run build and tested it using npm link. In addition, if the Admin SDK supports it, we could also make direct API requests as an alternative.

API reference: https://firebase.google.com/docs/firestore/reference/rest/v1beta1/projects.databases.documents/batchGet

boraberkcetintas committed:
Merge branch 'master' into feat-mcp-firestore-select-field

boraberkcetintas committed:
refactor(firestore): Use server-side filtering in get_documents tool

Implements the suggested enhancement to the get_documents tool by leveraging Firestore's native server-side filtering.

Previously, the tool fetched the entire document and filtered the fields within the firebase-tools server. This change pushes the field selection down to the Firestore backend using the DocumentMask (select) feature in the batchGet API call.

The corresponding unit tests have been updated to verify that the tool passes the select parameter to the underlying GCP client.

boraberkcetintas committed:
Merge branch 'master' into feat-mcp-firestore-select-field

joehan committed:
Merge branch 'master' into feat-mcp-firestore-select-field

joehan commented:
Looks like you just need to run 'npm run format' to fix the linter issues, and then this LGTM
